p2p/ipld/nmt_wrapper.go

```diff
@@ -0,0 +1,91 @@
+package ipld
+
+import (
+	"crypto/sha256"
+
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/nmt"
+	"github.com/lazyledger/nmt/namespace"
+	"github.com/lazyledger/rsmt2d"
+)
+
+// Fulfills the rsmt2d.Tree interface and rsmt2d.TreeConstructorFn function
+var _ rsmt2d.TreeConstructorFn = ErasuredNamespacedMerkleTree{}.Constructor
+var _ rsmt2d.Tree = &ErasuredNamespacedMerkleTree{}
+
+// ErasuredNamespacedMerkleTree wraps NamespaceMerkleTree to conform to the
+// rsmt2d.Tree interface while catering specifically to erasure data. For the
+// first half of the tree, it uses the first DefaultNamespaceIDLen number of
+// bytes of the data pushed to determine the namespace. For the second half, it
+// uses the parity namespace ID
+type ErasuredNamespacedMerkleTree struct {
+	squareSize uint64
+	pushCount  uint64
+	options    []nmt.Option
+	tree       *nmt.NamespacedMerkleTree
+}
+
+// NewErasuredNamespacedMerkleTree issues a new ErasuredNamespacedMerkleTree
+func NewErasuredNamespacedMerkleTree(squareSize uint64, setters ...nmt.Option) ErasuredNamespacedMerkleTree {
+	return ErasuredNamespacedMerkleTree{squareSize: squareSize, options: setters}
+}
+
+// Constructor acts as the rsmt2d.TreeConstructorFn for
+// ErasuredNamespacedMerkleTree
+func (w ErasuredNamespacedMerkleTree) Constructor() rsmt2d.Tree {
+	w.tree = nmt.New(sha256.New(), w.options...)
+	return &w
+}
+
+// Push adds the provided data to the underlying NamespaceMerkleTree, and
+// automatically uses the first DefaultNamespaceIDLen number of bytes as the
+// namespace unless the data pushed to the second half of the tree. Fulfills the
+// rsmt.Tree interface. NOTE: panics if there's an error pushing to underlying
+// NamespaceMerkleTree or if the tree size is exceeded
+func (w *ErasuredNamespacedMerkleTree) Push(data []byte) {
+	// determine the namespace based on where in the tree we're pushing
+	nsID := make(namespace.ID, types.NamespaceSize)
+
+	switch {
+	// panic if the tree size is exceeded
+	case w.pushCount > 2*w.squareSize:
+		panic("tree size exceeded")
+
+	// if the namespace is included in the data, use that ns
+	case w.pushCount+1 <= w.squareSize/2:
+		copy(nsID, data[:types.NamespaceSize])
+
+	// if the data is erasure data use the parity ns
+	default:
+		copy(nsID, types.ParitySharesNamespaceID)
```

Comment:
Note that when https://github.com/lazyledger/nmt/pull/27 lands, and if we attach the parity nID when calling Push, then this method can be simplified to just calling push on the underlying NMT, right?

Reply:
Unfortunately we will still need a wrapper, because `rsmt2d.RepairExtendedDatasquare` is not aware of how we erasure namespaces and `NamespacedMerkleTree` would always assume that the first 8 bytes is the namespace. Now that we're erasuring namespaces, the first 8 bytes of the the 2nd half of the tree is not a namespace, it's just erasure data. `nmt.Push` would panic.

Reply:
> Now that we're erasuring namespaces, the first 8 bytes of the the 2nd half of the tree is not a namespace,

I might still be missing something but why can't the caller of the NMT do the "preprocessing" and add the parity namespace in front of the parity shares (basically the same thing the wrapper currently does)?

Reply:
We can do that in lazyledger-core, but we would also have to do that in `rsmt2d.RepairDataSquare'.

Reply:
Ah, I see now. Thanks 🙏